README.rst
pytabix
=======

This module allows fast random access to sorted files compressed with bgzip_ and
indexed by tabix_. It includes a C extension with code from klib_. The bgzip
and tabix programs are available here_.

Alternatives: 

- Instead of using this module, you can install tabix_ yourself and
  easily call it from Python as shown at the bottom of this page.
  
- pysam_ - A python module for reading and manipulating Samfiles.
  It's a lightweight wrapper of the samtools C-API.
  Pysam also includes an interface for tabix_.

- hts-python_ - A pythonic wrapper for htslib_ C-API using python cffi_.

.. _pysam: https://github.com/pysam-developers/pysam
.. _hts-python: https://github.com/brentp/hts-python
.. _htslib: https://github.com/samtools/htslib
.. _cffi: https://cffi.readthedocs.org


Synopsis
--------

Open a local or remote file that has already been sorted, compressed, and indexed:

.. code:: python

    import tabix

    url = "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20100804/"
    url += "ALL.2of4intersection.20100804.genotypes.vcf.gz"

    tb = tabix.open(url)

    records = tb.query("1", 1000000, 1250000)
    records = tb.queryi(0, 1000000, 1250000)
    records = tb.querys("1:1000000-1250000")

    for record in records:
        print record[:3]

.. code:: python

    ['1', '1000760', 'rs75316104']
    ['1', '1000760', 'rs75316104']
    ['1', '1000894', 'rs114006445']
    ['1', '1000910', 'rs79750022']
    ['1', '1001177', 'rs4970401']
    ['1', '1001256', 'rs78650406']


Install
-------

.. code:: bash

    pip install --user pytabix

or

.. code:: bash

    wget https://pypi.python.org/packages/source/p/pytabix/pytabix-0.1.tar.gz
    tar xf pytabix-0.1.tar.gz
    cd pytabix-0.1
    python setup.py install --user

How to prepare a file for `tabix`
---------------------------------

1. Sort_ the file by chromosome, start, and end.
2. Compress the file with bgzip_.
3. Index the compressed file with tabix_.

.. code:: bash

    $ head -n5 example.bed | column -t
    chr19  53611131   53636172   ZNF415
    chr10  72149121   72150375   CEP57L1P1
    chr4   185009858  185139113  ENPP6
    chrX   132669772  133119672  GPC3
    chr6   134924279  134925376  FAM8A6P

1. Sort_ it by chromosome, start, and end.
2. Then, use bgzip_ to deflate the file into compressed blocks:

.. code:: bash

    $ sort -k1V -k2n -k3n example.bed | bgzip > example.bed.gz

3. Index the compressed file with tabix_:

.. code:: bash

    $ tabix -s 1 -b 2 -e 3 example.bed.gz
    
    $ ls
    example.bed  example.bed.gz  example.bed.gz.tbi


Alternative: Use `subprocess`
-----------------------------

.. code:: python

    from subprocess import Popen, PIPE
    
    def bgzip(filename):
        """Call bgzip to compress a file."""
        Popen(['bgzip', '-f', filename])
    
    def tabix_index(filename,
            preset="gff", chrom=1, start=4, end=5, skip=0, comment="#"):
        """Call tabix to create an index for a bgzip-compressed file."""
        Popen(['tabix', '-p', preset, '-s', chrom, '-b', start, '-e', end,
            '-S', skip, '-c', comment])
    
    def tabix_query(filename, chrom, start, end):
        """Call tabix and generate an array of strings for each line it returns."""
        query = '{}:{}-{}'.format(chrom, start, end)
        process = Popen(['tabix', '-f', filename, query], stdout=PIPE)
        for line in process.stdout:
            yield line.strip().split()


.. _bgzip: http://samtools.sourceforge.net/tabix.shtml
.. _tabix: http://samtools.sourceforge.net/tabix.shtml
.. _klib: https://github.com/jmarshall/klib
.. _here: http://sourceforge.net/projects/samtools/files/tabix/
.. _Sort: https://www.gnu.org/software/coreutils/manual/html_node/Details-about-version-sort.html#Details-about-version-sort


CHANGELOG.md
# Changelog

## 0.0.2 - 2015-03-21

### Fixed

-   `query` and `queryi` were incorrect. To get the same result as `querys`,
    the user had to subtract 1 from the start position. See:
    https://github.com/slowkow/pytabix/issues/5

-   The test for this module was not deterministic. A random test was
    generated, and the off-by-one bug in issue 5 was not caught because of
    this. Now I have a deterministic test that catches issue 5.

### Changed

-   The first version was called '0.1'. This does not follow semantic
    versioning standards. Now I've moved to '0.0.2' (MAJOR.MINOR.PATCH).
    See: http://semver.org/

## 0.1 - 2014-11-07

### Changed

-   Initial release.



test/test.py
#!/usr/bin/env python
"""
test_tabix.py
Hyeshik Chang <hyeshik@snu.ac.kr>
Kamil Slowikowski <slowikow@broadinstitute.org>
March 21, 2015

The MIT License

Copyright (c) 2011 Seoul National University.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""


import unittest
import random
import gzip
import tabix

TEST_FILE = 'test/example.gtf.gz'


def read_gtf(filename):
    """Read a GTF file and return a list of [chrom, start, end] items."""
    retval = []
    for line in gzip.open(filename):
        fields = line.rstrip().split('\t')
        chrom = fields[0]
        start = fields[3]
        end = fields[4]
        retval.append([chrom, start, end])
    return retval


def overlap1(a0, a1, b0, b1):
    """Check if two 1-based intervals overlap."""
    return int(a0) <= int(b1) and int(a1) >= int(b0)


def get_result(regions, chrom, start, end):
    retval = []
    for r in regions:
        if r[0] == chrom and overlap1(r[1], r[2], start, end):
            retval.append(r)
    return retval


class TabixTest(unittest.TestCase):
    regions = read_gtf(TEST_FILE)
    chrom = 'chr1'
    start = 25944
    end = 27000
    result = get_result(regions, chrom, start, end)
    tb = tabix.open(TEST_FILE)

    def test_query(self):
        it = self.tb.query(self.chrom, self.start, self.end)
        tb_result = [ [x[0], x[3], x[4]] for x in it ]
        self.assertEqual(self.result, tb_result)

    def test_querys(self):
        query = '{}:{}-{}'.format(self.chrom, self.start, self.end)
        it = self.tb.querys(query)
        tb_result = [ [x[0], x[3], x[4]] for x in it ]
        self.assertEqual(self.result, tb_result)

    def test_remote_file(self):
        file1 = "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20100804/" \
                "ALL.2of4intersection.20100804.genotypes.vcf.gz"
        tabix.open(file1)

    def test_remote_file_bad_url(self):
        file1 = "ftp://badurl"
        with self.assertRaises(tabix.TabixError):
            tabix.open(file1)


if __name__ == '__main__':
    unittest.main()


src/tabixmodule.c
/*-
 * The MIT License
 *
 * Copyright (c) 2011 by Seoul National University.
 *               2014 by Kamil Slowikowski <slowikow@broadinstitute.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/*
 * Contact: Hyeshik Chang <hyeshik@snu.ac.kr>
 */

#define PY_SSIZE_T_CLEAN
#include "Python.h"
#include "tabix.h"

static PyObject *TabixError;

typedef struct {
    PyObject_HEAD
    tabix_t *tb;
    char *fn;
} TabixObject;

typedef struct {
    PyObject_HEAD
    TabixObject *tbobj;
    ti_iter_t iter;
} TabixIteratorObject;

static PyTypeObject Tabix_Type, TabixIterator_Type;

/* --- TabixIterator --------------------------------------------------- */

static PyObject *
tabixiter_create(TabixObject *parentidx, ti_iter_t iter)
{
    TabixIteratorObject *self;

    self = PyObject_New(TabixIteratorObject, &TabixIterator_Type);
    if (self == NULL)
        return NULL;

    Py_INCREF(parentidx);
    self->tbobj = parentidx;
    self->iter = iter;

    return (PyObject *)self;
}

static void
tabixiter_dealloc(TabixIteratorObject *self)
{
    ti_iter_destroy(self->iter);
    Py_DECREF(self->tbobj);
    PyObject_Del(self);
}

static PyObject *
tabixiter_iter(PyObject *self)
{
    Py_INCREF(self);
    return self;
}

#if PY_MAJOR_VERSION < 3
# define PYOBJECT_FROM_STRING_AND_SIZE PyString_FromStringAndSize
#else
# define PYOBJECT_FROM_STRING_AND_SIZE PyUnicode_FromStringAndSize
#endif

static PyObject *
tabixiter_iternext(TabixIteratorObject *self)
{
    const char *chunk;
    int len, i;

    chunk = ti_read(self->tbobj->tb, self->iter, &len);
    if (chunk != NULL) {
        PyObject *ret, *column;
        Py_ssize_t colidx;
        const char *ptr, *begin;

        ret = PyList_New(0);
        if (ret == NULL)
            return NULL;

        colidx = 0;
        ptr = begin = chunk;
        for (i = len; i > 0; i--, ptr++)
            if (*ptr == '\t') {
                column = PYOBJECT_FROM_STRING_AND_SIZE(begin,
                                                       (Py_ssize_t)(ptr - begin));
                if (column == NULL || PyList_Append(ret, column) == -1) {
                    Py_DECREF(ret);
                    return NULL;
                }

                Py_DECREF(column);
                begin = ptr + 1;
                colidx++;
            }

        column = PYOBJECT_FROM_STRING_AND_SIZE(begin, (Py_ssize_t)(ptr - begin));
        if (column == NULL || PyList_Append(ret, column) == -1) {
            Py_DECREF(ret);
            return NULL;
        }
        Py_DECREF(column);

        return ret;
    }
    else
        return NULL;
}

static PyMethodDef tabixiter_methods[] = {
    {NULL, NULL} /* sentinel */
};

static PyTypeObject TabixIterator_Type = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "tabix.iter",      /*tp_name*/
    sizeof(TabixIteratorObject), /*tp_basicsize*/
    0,                          /*tp_itemsize*/
    /* methods */
    (destructor)tabixiter_dealloc,  /*tp_dealloc*/
    0,                          /*tp_print*/
    0,                          /*tp_getattr*/
    0,                          /*tp_setattr*/
    0,                          /*tp_compare*/
    0,                          /*tp_repr*/
    0,                          /*tp_as_number*/
    0,                          /*tp_as_sequence*/
    0,                          /*tp_as_mapping*/
    0,                          /*tp_hash*/
    0,                          /*tp_call*/
    0,                          /*tp_str*/
    0,                          /*tp_getattro*/
    0,                          /*tp_setattro*/
    0,                          /*tp_as_buffer*/
    Py_TPFLAGS_DEFAULT,         /*tp_flags*/
    "An iterator for records returned by tabix.query().\n\n"
    "    >>> tb = tabix.open(\"http://localhost/file.bgz\")\n"
    "    >>> records = list(tb.query(\"1\", 1000, 2000))\n",                          /*tp_doc*/
    0,                          /*tp_traverse*/
    0,                          /*tp_clear*/
    0,                          /*tp_richcompare*/
    0,                          /*tp_weaklistoffset*/
    tabixiter_iter,             /*tp_iter*/
    (iternextfunc)tabixiter_iternext, /*tp_iternext*/
    tabixiter_methods,          /*tp_methods*/
    0,                          /*tp_members*/
    0,                          /*tp_getset*/
    0,                          /*tp_base*/
    0,                          /*tp_dict*/
    0,                          /*tp_descr_get*/
    0,                          /*tp_descr_set*/
    0,                          /*tp_dictoffset*/
    0,                          /*tp_init*/
    0,                          /*tp_alloc*/
    0,                          /*tp_new*/
    0,                          /*tp_free*/
    0,                          /*tp_is_gc*/
};


/* --- Tabix ----------------------------------------------------------- */

static PyObject *
tabix_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
{
    TabixObject *self;
    const char *fn, *fnidx=NULL;
    static char *kwnames[]={"fn", "fnidx", NULL};
    tabix_t *tb;

    if (!PyArg_ParseTupleAndKeywords(args, kwds, "s|z:open",
                                     kwnames, &fn, &fnidx))
        return NULL;

    tb = ti_open(fn, fnidx);
    if (tb == NULL) {
        PyErr_SetString(TabixError, "Can't open the index file.");
        return NULL;
    }

    self = (TabixObject *)type->tp_alloc(type, 0);
    if (self == NULL)
        return NULL;

    self->tb = tb;
    self->fn = strdup(fn);

    return (PyObject *)self;
}

static void
tabix_dealloc(TabixObject *self)
{
    free(self->fn);
    ti_close(self->tb);
    PyObject_Del(self);
}

static PyObject *
tabix_query(TabixObject *self, PyObject *args)
{
    char *name;
    int begin, end;
    ti_iter_t result;

    if (!PyArg_ParseTuple(args, "sii:query", &name, &begin, &end))
        return NULL;

    result = ti_query(self->tb, name, begin, end);
    if (result == NULL) {
        PyErr_SetString(TabixError, "query failed");
        return NULL;
    }

    return tabixiter_create(self, result);
}

static PyObject *
tabix_queryi(TabixObject *self, PyObject *args)
{
    int tid, begin, end;
    ti_iter_t result;

    if (!PyArg_ParseTuple(args, "iii:queryi", &tid, &begin, &end))
        return NULL;

    result = ti_queryi(self->tb, tid, begin, end);
    if (result == NULL) {
        PyErr_SetString(TabixError, "query failed");
        return NULL;
    }

    return tabixiter_create(self, result);
}

static PyObject *
tabix_querys(TabixObject *self, PyObject *args)
{
    const char *reg;
    ti_iter_t result;

    if (!PyArg_ParseTuple(args, "s:querys", &reg))
        return NULL;

    result = ti_querys(self->tb, reg);
    if (result == NULL) {
        PyErr_SetString(TabixError, "query failed");
        return NULL;
    }

    return tabixiter_create(self, result);
}

static PyObject *
tabix_repr(TabixObject *self)
{
#if PY_MAJOR_VERSION < 3
    return PyString_FromFormat("<tabix fn=\"%s\">", self->fn);
#else
    return PyUnicode_FromFormat("<tabix fn=\"%s\">", self->fn);
#endif
}

static PyMethodDef tabix_methods[] = {
    {
        "query",
        (PyCFunction)tabix_query,
        METH_VARARGS,
        PyDoc_STR("Retrieve items within a region.\n\n"
                  "    >>> tb.query(\"chr1\", 1000, 2000)\n"
                  "    <tabix.iter at 0x17b86e50>\n\n"
                  "Parameters\n"
                  "----------\n"
                  "name : str\n"
                  "    Name of the sequence in the file.\n"
                  "start : int\n"
                  "    Start of the query region.\n"
                  "end : int\n"
                  "    End of the query region.\n")
    },
    {
        "queryi",
        (PyCFunction)tabix_queryi,
        METH_VARARGS,
        PyDoc_STR("Retrieve items within a region.\n\n"
                  "    >>> tb.queryi(0, 1000, 2000)\n"
                  "    <tabix.iter at 0x17b86e50>\n\n"
                  "Parameters\n"
                  "----------\n"
                  "tid : int\n"
                  "    Index of the sequence in the file (first is 0).\n"
                  "start : int\n"
                  "    Start of the query region.\n"
                  "end : int\n"
                  "    End of the query region.\n")
    },
    {
        "querys",
        (PyCFunction)tabix_querys,
        METH_VARARGS,
        PyDoc_STR("Retrieve items within a region.\n\n"
                  "    >>> tb.querys(\"chr1:1000-2000\")\n"
                  "    <tabix.iter at 0x17b86e50>\n\n"
                  "Parameters\n"
                  "----------\n"
                  "region : str\n"
                  "    Query string like \"seq:start-end\".\n")
    },
    /*
    {
        "header",
        (PyCFunction)tabix_header,
        METH_VARARGS,
        PyDoc_STR("Get the header for a file (VCF, SAM, GTF, etc.).\n\n"
                  "    >>> tb.header()\n")
    },
    */
    {NULL, NULL}           /* sentinel */
};
/*
	ti_iter_t ti_query(tabix_t *t, const char *name, int beg, int end);
	ti_iter_t ti_queryi(tabix_t *t, int tid, int beg, int end);
	ti_iter_t ti_querys(tabix_t *t, const char *reg);
*/

static PyTypeObject Tabix_Type = {
    /* The ob_type field must be initialized in the module init function
     * to be portable to Windows without using C++. */
    PyVarObject_HEAD_INIT(NULL, 0)
    "tabix.open",              /*tp_name*/
    sizeof(TabixObject),        /*tp_basicsize*/
    0,                          /*tp_itemsize*/
    /* methods */
    (destructor)tabix_dealloc,  /*tp_dealloc*/
    0,                          /*tp_print*/
    0,                          /*tp_getattr*/
    0,                          /*tp_setattr*/
    0,                          /*tp_compare*/
    (reprfunc)tabix_repr,       /*tp_repr*/
    0,                          /*tp_as_number*/
    0,                          /*tp_as_sequence*/
    0,                          /*tp_as_mapping*/
    0,                          /*tp_hash*/
    0,                          /*tp_call*/
    0,                          /*tp_str*/
    0,                          /*tp_getattro*/
    0,                          /*tp_setattro*/
    0,                          /*tp_as_buffer*/
    Py_TPFLAGS_DEFAULT,         /*tp_flags*/
    "Open a local or remote file compressed with bgzip and indexed with tabix.\n\n"
    "    >>> tb = tabix.open(\"http://localhost/file.bgz\")\n"
    "    >>> tb.query(\"1\", 1000, 2000)\n"
    "    <tabix.iter at 0x17b86e50>",                          /*tp_doc*/
    0,                          /*tp_traverse*/
    0,                          /*tp_clear*/
    0,                          /*tp_richcompare*/
    0,                          /*tp_weaklistoffset*/
    0,                          /*tp_iter*/
    0,                          /*tp_iternext*/
    tabix_methods,              /*tp_methods*/
    0,                          /*tp_members*/
    0,                          /*tp_getset*/
    0,                          /*tp_base*/
    0,                          /*tp_dict*/
    0,                          /*tp_descr_get*/
    0,                          /*tp_descr_set*/
    0,                          /*tp_dictoffset*/
    0,                          /*tp_init*/
    0,                          /*tp_alloc*/
    (newfunc)tabix_new,         /*tp_new*/
    0,                          /*tp_free*/
    0,                          /*tp_is_gc*/
};
/* --------------------------------------------------------------------- */

static PyMethodDef tabix_functions[] = {
    {NULL, NULL} /* sentinel */
};

PyDoc_STRVAR(module_doc,
"This module allows fast random access to files compressed with ``bgzip`` and\n"
"indexed by ``tabix``. It includes a C extension with code from klib_.\n"
"\n"
"Genomics data is often in a table where each row corresponds to a genomic\n"
"region::\n"
"\n"
"    chrom  pos      snp\n"
"    1      1000760  rs75316104\n"
"    1      1000894  rs114006445\n"
"    1      1000910  rs79750022\n"
"    1      1001177  rs4970401\n"
"    1      1001256  rs78650406\n"
"\n"
"With ``tabix``, you can quickly retrieve all rows in a genomic region by\n"
"specifying a query with a sequence name, start, and end::\n"
"\n"
"    import tabix\n"
"\n"
"    # Open a remote or local file.\n"
"    url = \"ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20100804/\"\n"
"    url += \"ALL.2of4intersection.20100804.genotypes.vcf.gz\"\n"
"\n"
"    tb = tabix.open(url)\n"
"\n"
"    # These queries are identical. A query returns an iterator over the\n"
"    # results.\n"
"    records = tb.query(\"1\", 1000000, 1250000)\n"
"\n"
"    records = tb.queryi(0, 1000000, 1250000)\n"
"\n"
"    records = tb.querys(\"1:1000000-1250000\")\n"
"\n"
"    # Each record is a list of strings.\n"
"    for record in records:\n"
"        print record[:5]\n"
"            break\n"
"\n"
"The ``bgzip`` and ``tabix`` programs are provided in samtools_.\n"
"\n"
".. _klib: https://github.com/jmarshall/klib\n"
".. _samtools: http://samtools.sourceforge.net\n");

#if PY_MAJOR_VERSION >= 3
static struct PyModuleDef tabixmodule = { 
    PyModuleDef_HEAD_INIT,
    "tabix",
    module_doc,
    -1, 
    tabix_functions,
    NULL,
    NULL,
    NULL,
    NULL
};
#endif

#if PY_MAJOR_VERSION < 3
PyMODINIT_FUNC inittabix(void)
#else
PyMODINIT_FUNC PyInit_tabix(void)
#endif
{
    PyObject *m;

    if (PyType_Ready(&Tabix_Type) < 0)
        goto fail;
    if (PyType_Ready(&TabixIterator_Type) < 0)
        goto fail;

#if PY_MAJOR_VERSION < 3
    m = Py_InitModule3("tabix", tabix_functions, module_doc);
#else
    m = PyModule_Create(&tabixmodule);
#endif
    if (m == NULL)
        goto fail;

    if (TabixError == NULL) {
        TabixError = PyErr_NewException("tabix.TabixError", NULL, NULL);
        if (TabixError == NULL)
            goto fail;
    }
    Py_INCREF(TabixError);
    PyModule_AddObject(m, "TabixError", TabixError);

    PyModule_AddObject(m, "open", (PyObject *)&Tabix_Type);
    PyModule_AddObject(m, "iter", (PyObject *)&TabixIterator_Type);

#if PY_MAJOR_VERSION >= 3
    return m;
#endif

 fail:
#if PY_MAJOR_VERSION < 3
    return;
#else
    return NULL;
#endif
}
